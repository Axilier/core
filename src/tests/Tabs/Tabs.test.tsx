/** @format */

import React from 'react';
import { mount, shallow } from 'enzyme';
import { Tab, Tabs } from '../../index';
import { expectToThrow } from '../expectToThrow';

const onChange = jest.fn();
const tabMenu = mount(
    <Tabs orientation={'row'}>
        <Tab>test1</Tab>
        <Tab>test2</Tab>
        <Tab>test3</Tab>
    </Tabs>,
);

describe('when the `onclick` prop is set', () => {
    beforeAll(() => tabMenu.setProps({ onChange }));
    it('should be called whenever a different tab is selected', () => {
        tabMenu.childAt(0).childAt(0).childAt(0).simulate('click');
        expect(onChange).toHaveBeenCalledWith(0);
        tabMenu.childAt(0).childAt(0).childAt(1).simulate('click');
        expect(onChange).toHaveBeenCalledWith(1);
    });
    afterAll(() => jest.clearAllMocks());
});
describe('when the `onclick` prop is not set', () => {
    beforeAll(() => tabMenu.setProps({ onChange: undefined }));
    it('should be called whenever a different tab is selected', () => {
        tabMenu.childAt(0).childAt(0).childAt(0).simulate('click');
        expect(onChange).not.toHaveBeenCalled();
        tabMenu.childAt(0).childAt(0).childAt(1).simulate('click');
        expect(onChange).not.toHaveBeenCalled();
    });
});
describe('when I provide children of a type not Tab', () => {
    it('should throw an error', () => {
        expectToThrow(
            () =>
                mount(
                    <Tabs orientation={'row'}>
                        <div />
                    </Tabs>,
                ),
            'children of TabMenu must be of type Tab',
        );
    });
});
describe('when I render the component', () => {
    it('should render as expected', () => {
        expect(tabMenu).toMatchSnapshot();
    });
    it('should have a selected indicator', () => {
        expect(tabMenu.childAt(0).childAt(0).childAt(3).is('#indicator')).toBe(
            true,
        );
    });
    it('should have an indicator with a width 1/3 of the tabMenu', async () => {
        expect(tabMenu.find('#indicator').props().style).toHaveProperty(
            'width',
            'calc( 100% / 3)',
        );
    });
    describe('when the `tabNotSelectedColor` prop is not set', () => {
        it('should set the non selected tabs background colour to transparent', () => {
            expect(tabMenu.childAt(0).props().style).toHaveProperty(
                'backgroundColor',
                'transparent',
            );
        });
    });
    describe('when the `tabNotSelectedColor` prop is set', () => {
        beforeAll(() => tabMenu.setProps({ tabNotSelectedColor: 'red' }));
        it('should set the non selected tabs background colour to the desired colour', () => {
            expect(tabMenu.childAt(0).props().style).toHaveProperty(
                'backgroundColor',
                'red',
            );
        });
        afterAll(() => tabMenu.setProps({ tabNotSelectedColor: undefined }));
    });
    describe('when the `tabIndicatorColor` prop is not set', () => {
        it('should set the indicator bar to transparent', () => {
            expect(
                tabMenu.childAt(0).childAt(0).childAt(3).props().style,
            ).toHaveProperty('backgroundColor', '#057AFF');
        });
    });
    describe('when the `tabIndicatorColor` prop is set', () => {
        beforeAll(() => tabMenu.setProps({ tabIndicatorColor: 'red' }));
        it('should set the indicator bar to the desired colour ', () => {
            expect(
                tabMenu.childAt(0).childAt(0).childAt(3).props().style,
            ).toHaveProperty('backgroundColor', 'red');
        });
        afterAll(() => tabMenu.setProps({ tabIndicatorColor: undefined }));
    });
    describe('when the `tabSelectedColor` prop is not set', () => {
        it('should set the indicator bar to transparent', () => {
            expect(
                tabMenu.childAt(0).childAt(0).childAt(4).props().style,
            ).toHaveProperty('backgroundColor', '#F4F4F4');
        });
    });
    describe('when the `tabSelectedColor` prop is set', () => {
        beforeAll(() => tabMenu.setProps({ tabSelectedColor: 'red' }));
        it('should set the indicator bar to transparent', () => {
            expect(
                tabMenu.childAt(0).childAt(0).childAt(4).props().style,
            ).toHaveProperty('backgroundColor', 'red');
        });
        afterAll(() => tabMenu.setProps({ tabSelectedColor: undefined }));
    });
    describe('when the orientation is row', () => {
        it('should set the width of the indicator to a calculation', () => {
            expect(
                tabMenu.childAt(0).childAt(0).childAt(3).props().style,
            ).toHaveProperty('width', 'calc( 100% / 3)');
        });
        it('should set the height of the indicator to 2px', () => {
            expect(
                tabMenu.childAt(0).childAt(0).childAt(3).props().style,
            ).toHaveProperty('height', '2px');
        });
    });
    describe('when the orientation is column', () => {
        beforeAll(() => tabMenu.setProps({ orientation: 'column' }));
        it('should set the width of the indicator to 2px', () => {
            expect(
                tabMenu.childAt(0).childAt(0).childAt(3).props().style,
            ).toHaveProperty('width', '2px');
        });
        it('should set the width of the indicator to 2px', () => {
            expect(
                tabMenu.childAt(0).childAt(0).childAt(3).props().style,
            ).toHaveProperty('height', 'calc( 100% / 3)');
        });
        afterAll(() => tabMenu.setProps({ orientation: 'row' }));
    });
});
describe('when I provide 3 children of type Tab', () => {
    it('should show the three Tab children', () => {
        expect(tabMenu.childAt(0).childAt(0).children()).toHaveLength(5);
    });
});
