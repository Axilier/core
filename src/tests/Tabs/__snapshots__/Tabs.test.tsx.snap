// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when I render the component should render as expected 1`] = `
<Tabs
  orientation="row"
>
  <Layout
    className="tabMenu"
    elemRef={
      Object {
        "current": <div
          class="tabMenu"
          style="display: flex; flex-direction: row; max-height: 120px; width: 100%; color: black; background-color: transparent;"
        >
          <div
            class="tab tabHorizontal"
            style="display: flex; flex-direction: row; align-items: center; justify-content: center;"
          >
            test1
            <div
              class="tabSelectedShadow horizontalShadow"
              style="display: block;"
            />
          </div>
          <div
            class="tab tabHorizontal"
            style="display: flex; flex-direction: row; align-items: center; justify-content: center;"
          >
            test2
            <div
              class="tabSelectedShadow horizontalShadow"
              style="display: block;"
            />
          </div>
          <div
            class="tab tabHorizontal"
            style="display: flex; flex-direction: row; align-items: center; justify-content: center;"
          >
            test3
            <div
              class="tabSelectedShadow horizontalShadow"
              style="display: block;"
            />
          </div>
          <div
            class="tabIndicator"
            id="indicator"
            style="background-color: rgb(5, 122, 255); width: calc( 100% / 3); height: 2px; top: calc( 100% - 2px ); left: 0px;"
          />
          <div
            class="tabSelectedBackground"
            id="selectedTab"
            style="height: 100%; width: 0px; top: 0px; left: 0px; background-color: rgb(244, 244, 244);"
          />
        </div>,
      }
    }
    orientation="row"
    style={
      Object {
        "backgroundColor": "transparent",
        "color": "black",
        "maxHeight": "120px",
        "width": "100%",
      }
    }
  >
    <div
      className="tabMenu"
      style={
        Object {
          "backgroundColor": "transparent",
          "color": "black",
          "display": "flex",
          "flexDirection": "row",
          "maxHeight": "120px",
          "width": "100%",
        }
      }
    >
      <Tab
        key="test1-loop"
        onClick={[Function]}
        orientation="row"
        showNotSelectedShadow={true}
      >
        <Layout
          center={true}
          className="tab tabHorizontal"
          onClick={[Function]}
          orientation="row"
        >
          <div
            className="tab tabHorizontal"
            onClick={[Function]}
            style={
              Object {
                "alignItems": "center",
                "display": "flex",
                "flexDirection": "row",
                "justifyContent": "center",
              }
            }
          >
            test1
            <div
              className="tabSelectedShadow horizontalShadow"
              style={
                Object {
                  "display": "block",
                }
              }
            />
          </div>
        </Layout>
      </Tab>
      <Tab
        key="test2-loop"
        onClick={[Function]}
        orientation="row"
        showNotSelectedShadow={true}
      >
        <Layout
          center={true}
          className="tab tabHorizontal"
          onClick={[Function]}
          orientation="row"
        >
          <div
            className="tab tabHorizontal"
            onClick={[Function]}
            style={
              Object {
                "alignItems": "center",
                "display": "flex",
                "flexDirection": "row",
                "justifyContent": "center",
              }
            }
          >
            test2
            <div
              className="tabSelectedShadow horizontalShadow"
              style={
                Object {
                  "display": "block",
                }
              }
            />
          </div>
        </Layout>
      </Tab>
      <Tab
        key="test3-loop"
        onClick={[Function]}
        orientation="row"
        showNotSelectedShadow={true}
      >
        <Layout
          center={true}
          className="tab tabHorizontal"
          onClick={[Function]}
          orientation="row"
        >
          <div
            className="tab tabHorizontal"
            onClick={[Function]}
            style={
              Object {
                "alignItems": "center",
                "display": "flex",
                "flexDirection": "row",
                "justifyContent": "center",
              }
            }
          >
            test3
            <div
              className="tabSelectedShadow horizontalShadow"
              style={
                Object {
                  "display": "block",
                }
              }
            />
          </div>
        </Layout>
      </Tab>
      <div
        className="tabIndicator"
        id="indicator"
        key="indicator"
        style={
          Object {
            "backgroundColor": "#057AFF",
            "height": "2px",
            "left": 0,
            "top": "calc( 100% - 2px )",
            "width": "calc( 100% / 3)",
          }
        }
      />
      <div
        className="tabSelectedBackground"
        id="selectedTab"
        key="selectedTab"
        style={
          Object {
            "backgroundColor": "#F4F4F4",
            "height": "100%",
            "left": 0,
            "top": 0,
            "width": 0,
          }
        }
      />
    </div>
  </Layout>
</Tabs>
`;
